lazyscribe/project.py

```diff
@@ -148,9 +173,37 @@ def save(self):
                     self[slug].last_updated_by = self.author
 
         data = list(self)
-        with open(self.fpath, "w") as outfile:
+        with self.fs.open(self.fpath, "w") as outfile:
             json.dump(data, outfile, sort_keys=True, indent=4)
 
+        for exp in self.experiments:

```

Comment:
Because we're not doing any input checking upon artifact initialization, you can conceivably run into a scenario where you initialize an artifact early in a workflow, get through the flow successfully, and then get an exception that your artifact couldn't be saved (e.g. not json serializable, etc.). The way this is set up, the project Jason gets saved before the artifacts are persisted so you could wind up with some artifacts listed in the json that don't actually exist on disk.

A couple options:
1. Reverse the order of operations, e.g. persist artifacts then save the project json.
2. Use a try/except block to clean up any artifacts that were unsuccessful in saving.

Reply:
I wonder if we need to clean up the artifacts after failure. I reversed the order of the saves, but if you encounter an error persisting your artifacts you can fix the issue with the object and just call `save` again, right?